source/h2_connection.c

```diff
@@ -2154,17 +2263,82 @@ static int s_connection_ping(
     }
 
     return AWS_OP_SUCCESS;
+
+closed:
+    CONNECTION_LOG(ERROR, connection, "Failed to send ping, connection is closed or closing.");
+    aws_h2_frame_destroy(ping_frame);
+    aws_mem_release(connection->base.alloc, pending_ping);
+    return aws_raise_error(AWS_ERROR_INVALID_STATE);
 }
 
-/* Send a GOAWAY with the lowest possible last-stream-id */
-static void s_send_goaway(struct aws_h2_connection *connection, enum aws_http2_error_code h2_error_code) {
+static int s_connection_send_goaway(
+    struct aws_http_connection *connection_base,
+    uint32_t http2_error,
+    bool allow_more_streams,
+    const struct aws_byte_cursor *optional_debug_data) {
+
+    struct aws_h2_connection *connection = AWS_CONTAINER_OF(connection_base, struct aws_h2_connection, base);
+    struct aws_h2_pending_goaway *pending_goaway =
+        s_new_pending_goaway(connection->base.alloc, http2_error, allow_more_streams, optional_debug_data);
+    if (!pending_goaway) {
+        return AWS_OP_ERR;
+    }
+
+    bool was_cross_thread_work_scheduled = false;
+    bool connection_open;
+    { /* BEGIN CRITICAL SECTION */
+        s_lock_synced_data(connection);
+
+        connection_open = connection->synced_data.is_open;
+        if (!connection_open) {
+            s_unlock_synced_data(connection);
+            goto closed;
+        }
+        was_cross_thread_work_scheduled = connection->synced_data.is_cross_thread_work_task_scheduled;
+        connection->synced_data.is_cross_thread_work_task_scheduled = true;
+        aws_linked_list_push_back(&connection->synced_data.pending_goaway_list, &pending_goaway->node);
+        s_unlock_synced_data(connection);
+    } /* END CRITICAL SECTION */
+
+    if (!was_cross_thread_work_scheduled) {
+        CONNECTION_LOG(TRACE, connection, "Scheduling cross-thread work task");
+        aws_channel_schedule_task_now(connection->base.channel_slot->channel, &connection->cross_thread_work_task);
+    }
+    return AWS_OP_SUCCESS;
+
+closed:
+    CONNECTION_LOG(ERROR, connection, "Failed to send goaway, connection is closed or closing.");
+    aws_mem_release(connection->base.alloc, pending_goaway);
+    return aws_raise_error(AWS_ERROR_INVALID_STATE);
+}
+
+/* Send a GOAWAY with the lowest possible last-stream-id or graceful shutdown warning */
+static void s_send_goaway(
+    struct aws_h2_connection *connection,
+    uint32_t h2_error_code,
+    bool allow_more_streams,
+    const struct aws_byte_cursor *optional_debug_data) {
     AWS_PRECONDITION(aws_channel_thread_is_callers_thread(connection->base.channel_slot->channel));
+    AWS_PRECONDITION(!allow_more_streams || (h2_error_code == AWS_HTTP2_ERR_NO_ERROR));
```

Comment:
I'm not sure this should be an assert.
It's WEIRD, but it's not against the law.
We could log about it at DEBUG or WARN level?

Reply:
mmm, that's the internal function. If this is called with this kind of error, probably something is wrong internally... In this case, would it be better to assert?

Reply:
And if user calls the API with this kind of error, I just ignore this kind of error when we create the pending_goaway structure. It's probably better to log it.
'''
pending_goaway->http2_error = allow_more_streams ? AWS_HTTP2_ERR_NO_ERROR : http2_error;
'''

Reply:
this is a pretty arbitrary call... but if it's not against the law, we should do what the user tells us to do. Let's just let them send a non-zero error code, even if allow_more_streams is true.

Reply:
Yeah, makes sense. Will do.